receiver/collectdreceiver/receiver.go

```diff
@@ -0,0 +1,182 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collectdreceiver
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	"go.uber.org/zap"
+)
+
+var (
+	errNilNextConsumer = errors.New("nil nextConsumer")
+	errAlreadyStarted  = errors.New("already started")
+	errAlreadyStopped  = errors.New("already stopped")
+)
+
+var _ receiver.MetricsReceiver = (*collectdReceiver)(nil)
+
+// collectdReceiver implements the receiver.TraceReceiver for CollectD protocol.
+type collectdReceiver struct {
+	sync.Mutex
+	logger             *zap.Logger
+	addr               string
+	server             *http.Server
+	defaultAttrsPrefix string
+	nextConsumer       consumer.MetricsConsumer
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+// New creates the CollectD receiver with the given parameters.
+func New(
+	logger *zap.Logger,
+	addr string,
+	timeout time.Duration,
+	defaultAttrsPrefix string,
+	nextConsumer consumer.MetricsConsumer) (receiver.MetricsReceiver, error) {
+	if nextConsumer == nil {
+		return nil, errNilNextConsumer
+	}
+
+	r := &collectdReceiver{
+		logger:             logger,
+		addr:               addr,
+		nextConsumer:       nextConsumer,
+		defaultAttrsPrefix: defaultAttrsPrefix,
+	}
+	r.server = &http.Server{
+		Addr:         addr,
+		Handler:      r,
+		ReadTimeout:  timeout,
+		WriteTimeout: timeout,
+	}
+	return r, nil
+}
+
+const metricsSource string = "CollectD"
+
+// TraceSource returns the name of the trace data source.
+func (cdr *collectdReceiver) MetricsSource() string {
+	return metricsSource
+}
+
+// StartMetricsReception starts an HTTP server that can process CollectD JSON requests.
+func (cdr *collectdReceiver) StartMetricsReception(host receiver.Host) error {
+	cdr.Lock()
+	defer cdr.Unlock()
+
+	err := errAlreadyStarted
+	cdr.startOnce.Do(func() {
+		err = nil
+		go func() {
+			_ = cdr.server.ListenAndServe()
+		}()
+	})
+
+	return err
+}
+
+// StopMetricsReception stops the CollectD receiver.
+func (cdr *collectdReceiver) StopMetricsReception() error {
+	cdr.Lock()
+	defer cdr.Unlock()
+
+	var err = errAlreadyStopped
+	cdr.stopOnce.Do(func() {
+		err = cdr.server.Shutdown(context.Background())
+	})
+	return err
+}
+
+// ServeHTTP acts as the default and only HTTP handler for the CollectD receiver.
+func (cdr *collectdReceiver) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	recordRequestReceived()
+
+	if r.Method != "POST" {
+		recordRequestErrors()
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		recordRequestErrors()
+		w.WriteHeader(http.StatusInternalServerError)
```

Comment:
I thought there might be reasons other than client sending bad data for this but changed to badrequest after reading source of net/http. Most cases should be because of bad body if at all.